Improve error handling

pepa65 opened the pull request:
Sourcing still exits, because `set -e` forces exits after errors. Errors should instead be caught and handled properly, because when the script is sourced, any error will terminate the login shell.

Not sure if you find adding setcap to the install desirable.
My aim is for this script to be usable for my own upgrades. For that to happen I will implement an upgrade that selects all the features that were selected before. BUT -- I only know how to select the Directives/Middleware features. Is there a way for the URL to select all the features, also the DNS server and DNS providers??

pepa65 committed:
Fix mistype . instead of , in comment

pepa65 commented:
By the way, I noticed that the file served at getcaddy.com is a previous version than current master...

mholt commented:
So what do you mean by sourcing exactly?

pepa65 commented:
The script recommends doing `curl https://getcaddy.com |bash` which is very similar to running the script as an executable. (One would have to fully trust the script to do what you want, and probably examine it before doing this.) People could also first want to download the script to examine (and perhaps modify?) it locally. Then it could be run in two ways: sourcing `source getcaddy.sh` or running it the same way as recommended `bash getcaddy.sh` or running it as an executable `chmod +x getcaddy.sh; ./getcaddy.sh`. All three ways could work. But if `set -e` is used, it will exit your shell on encountering any error condition, also on the outcome of certain commands, like not finding things with grep; it is not a good way to handle errors, but it does make sure the script stops executing (if that is mainly what you want). In the case of sourcing, people are also thrown out of their shell, and most people really don't like that.

mholt commented:
I see -- I forgot that `source` was a way that people run scripts... Thanks for clarifying!

mholt commented on index.txt:
```diff
@@ -22,15 +22,23 @@
+# Or it can be run after downloading:
+#
+# $ wget getcaddy.com -O getcaddy.sh
+# $ bash getcaddy.sh git,mailout
```
I'd like to keep these comments at the beginning as short as possible, and I think that people who are concerned about this already know to do this, so let's remove this.

mholt commented on index.txt:
```diff
@@ -22,15 +22,23 @@
+# The full list of available features is:
+#  awslambda,cors,expires,filemanager,filter,git,hugo,ipfilter,jsonp,jwt,
+#  locale,mailout,minify,multipass,prometheus,ratelimit,realip,search,upload
+#
```
Also take this list out.

mholt commented on index.txt:
```diff
@@ -162,13 +170,14 @@ install_caddy()
 
 	echo "Putting caddy in $install_path (may require password)"
 	$sudo_cmd mv "$PREFIX/tmp/$caddy_bin" "$install_path/$caddy_bin"
+	$sudo_cmd setcap cap_net_bind_service=+ep "$install_path/$caddy_bin"
```
Setcap may not be available on some systems. We should make sure it exists before running this. (e.g. doesn't exist on Mac)

mholt reviewed (changes requested):
This is a good start, but the PR is doing a lot of things at once. A few revisions and it should be good to go. Thanks!

pepa65 commented:
So would you like me to submit a pull request for doing setcap only on non-Mac and the little bugfixes?
Or would you want the whole set except the lines in the intro and introduce a check on setcap??
Happy to do both, but I'll start with the first.

mholt commented:
Typically, just one thing per PR is nice. In this case, if you want to just make the changes I've requested then I can merge this.

pepa65 committed:
Removed the requested lines from the intro
Tested for the presence of setcap binary
Fixed 2 bad bugs
- using literal caddy when it could be caddy.exe
- adding -- in rm, as the presence of a PREFIX variable with the wrong content
  could be bad (like: '-rf ')

pepa65 commented:
I don't know how to resolve the conflicts using this interface. It should be the latest commit of course...

mholt commented:
Nice work - thanks! Once the conflict is resolved I'll merge this in. :) Git has marked where the conflict is, you just have to edit the file so that the two conflicting changes are merged/resolved.

pepa65 commented:
The one change is just to invoke setcap, which is undesirable. Then I made it conditional by editing the file. I don't understand how I can edit the file to merge/resolve this. I think the earlier wrong change should just be erased.

mholt commented:
The easy way is to just click "Resolve conflicts", make the file look like it's supposed to look, and then click "Mark resolved".

pepa65 committed:
Merge branch 'master' into master

pepa65 commented:
In the current code, what is really not so good is that the caddy binary is being backed up before it is clear that it is ever going to be replaced. This backing up needs only to happen once we're ready to actually replace the binary, after downloading and unpacking the archive.

Can I give you a PR for that?

mholt commented:
@pepa65 Sure!
